File: parser.py
class ASTNode:
    pass

class Number(ASTNode):
    def __init__(self, value):
        self.value = value

    def __repr__(self):
        return f"Number({self.value})"

class Variable(ASTNode):
    def __init__(self, name):
        self.name = name

    def __repr__(self):
        return f"Variable({self.name})"

class Assign(ASTNode):
    def __init__(self, name, value):
        self.name = name
        self.value = value

    def __repr__(self):
        return f"Assign({self.name} := {self.value})"

class BinaryOp(ASTNode):
    def __init__(self, left, op, right):
        self.left = left
        self.op = op
        self.right = right

    def __repr__(self):
        return f"({self.left} {self.op} {self.right})"

class UnaryOp(ASTNode):  # 針對 `not` 運算
    def __init__(self, op, operand):
        self.op = op
        self.operand = operand

    def __repr__(self):
        return f"({self.op} {self.operand})"

class FunctionCall(ASTNode):
    def __init__(self, name, args):
        self.name = name
        self.args = args

    def __repr__(self):
        return f"{self.name}({', '.join(map(str, self.args))})"


class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0

    def eat(self, token_type):
        if self.pos < len(self.tokens) and self.tokens[self.pos][0] == token_type:
            self.pos += 1
        else:
            raise SyntaxError(f"Expected {token_type}, got {self.tokens[self.pos]}")

    def parse_expr(self):
        """解析 `+`, `-`, `&&`, `||` 運算"""
        left = self.parse_term()

        while self.pos < len(self.tokens) and self.tokens[self.pos][0] in ("PLUS", "MINUS", "AND", "OR"):
            op = self.tokens[self.pos][1]
            self.pos += 1
            right = self.parse_term()
            left = BinaryOp(left, op, right)

        return left

    
    def parse_term(self):
        """解析 `*`, `/` 以及 `!` 運算"""
        left = self.parse_factor()

        while self.pos < len(self.tokens) and self.tokens[self.pos][0] in ("MUL", "DIV"):
            op = self.tokens[self.pos][1]
            self.pos += 1
            right = self.parse_factor()
            left = BinaryOp(left, op, right)

        return left

    def parse_factor(self):
        """解析數字、變數，並支援 `!` 和 `-`"""
        token = self.tokens[self.pos]

        if token[0] == "NUMBER":
            self.pos += 1
            return Number(token[1])

        elif token[0] == "IDENT":
            self.pos += 1
            return Variable(token[1])

        elif token[0] == "NOT":  # `!` 運算
            op = token[1]
            self.pos += 1
            operand = self.parse_factor()
            return UnaryOp(op, operand)

        elif token[0] == "MINUS":  # **負數 `-10` 轉換為 `UnaryOp("-", Number(10))`**
            op = token[1]
            self.pos += 1
            operand = self.parse_factor()
            return UnaryOp(op, operand)

        else:
            raise SyntaxError(f"Unexpected token {token}")

    def parse_function_call(self, name):
        """解析函式呼叫，例如 `print(a, b, c)`"""
        self.eat("LPAREN")  # 跳過 `(`
        args = []
        while self.tokens[self.pos][0] != "RPAREN":
            args.append(self.parse_expr())  # 解析變數或數字
            if self.tokens[self.pos][0] == "COMMA":
                self.eat("COMMA")  # **跳過逗號**
            elif self.tokens[self.pos][0] == "RPAREN":
                break
            else:
                raise SyntaxError(f"Unexpected token {self.tokens[self.pos]} in function call")
        self.eat("RPAREN")  # 跳過 `)`
        return FunctionCall(name, args)


    def parse(self):
        ast = []
        while self.pos < len(self.tokens):
            token_type, token_value = self.tokens[self.pos]

            if token_type == "IDENT":
                name = token_value
                self.eat("IDENT")

                if self.tokens[self.pos][0] == "LPAREN":
                    ast.append(self.parse_function_call(name))
                else:
                    self.eat("ASSIGN")  # 這裡報錯，因為 `NEWLINE` 可能未被處理
                    value = self.parse_expr()
                    ast.append(Assign(name, value))

            elif token_type == "NEWLINE":
                self.eat("NEWLINE")  # **正確跳過換行**
                continue

            else:
                raise SyntaxError(f"Unexpected token {self.tokens[self.pos]}")

        return ast

File: src/parser.rs
use crate::ast::ASTNode;
use crate::error::{GluxError, glux_error};
use crate::lexer::Token;

pub struct Parser {
    tokens: Vec<Token>,
    pos: usize,
}

impl Parser {
    /// **建立一個新的 Parser**
    pub fn new(tokens: Vec<Token>) -> Self {
        Parser { tokens, pos: 0 }
    }

    /// **取得當前 token**（若超出範圍則回傳 `EOF`）
    fn current_token(&self) -> &Token {
        self.tokens.get(self.pos).unwrap_or(&Token::EOF)
    }

    /// **前進一個 token**
    fn advance(&mut self) {
        if self.pos < self.tokens.len() {
            self.pos += 1;
        }
    }

    /// **解析整個程式（目前只處理單一表達式）**
    pub fn parse(&mut self) -> Result<ASTNode, GluxError> {
        let expr = self.parse_expression()?;
        if let Token::EOF = self.current_token() {
            Ok(expr)
        } else {
            glux_error!(
                ParserError,
                "Unexpected token after expression: {:?}",
                self.current_token()
            )
        }
    }

    /// **解析表達式**（支援 `+`、`-`）
    pub fn parse_expression(&mut self) -> Result<ASTNode, GluxError> {
        let mut node = self.parse_term()?;
        while let Token::Operator(op) = self.current_token() {
            if op == "+" || op == "-" {
                let op_clone = op.clone();
                self.advance();
                let right = self.parse_term()?;
                node = ASTNode::BinaryOp {
                    op: op_clone,
                    left: Box::new(node),
                    right: Box::new(right),
                };
            } else {
                break;
            }
        }
        Ok(node)
    }

    /// **解析項（支援 `*`、`/`、`%`）**
    fn parse_term(&mut self) -> Result<ASTNode, GluxError> {
        let mut node = self.parse_factor()?;
        while let Token::Operator(op) = self.current_token() {
            if op == "*" || op == "/" || op == "%" {
                let op_clone = op.clone();
                self.advance();
                let right = self.parse_factor()?;
                node = ASTNode::BinaryOp {
                    op: op_clone,
                    left: Box::new(node),
                    right: Box::new(right),
                };
            } else {
                break;
            }
        }
        Ok(node)
    }

    /// **解析因子（數字、字串、識別符、括號、一元運算）**
    fn parse_factor(&mut self) -> Result<ASTNode, GluxError> {
        match self.current_token() {
            // **處理數字**
            Token::Number(n) => {
                let num = n.clone();
                self.advance();
                Ok(ASTNode::Number(num))
            }

            // **處理字串**
            Token::StringLiteral(s) => {
                let s_clone = s.clone();
                self.advance();
                Ok(ASTNode::StringLiteral(s_clone))
            }

            // **處理識別符（可能是變數或函數呼叫）**
            Token::Identifier(id) => {
                let id_clone = id.clone();
                self.advance();
                if let Token::Symbol('(') = self.current_token() {
                    self.parse_function_call(id_clone)
                } else {
                    Ok(ASTNode::Identifier(id_clone))
                }
            }

            // **處理括號 `(expr)`**
            Token::Symbol('(') => {
                self.advance();
                let expr = self.parse_expression()?;
                if let Token::Symbol(')') = self.current_token() {
                    self.advance();
                    Ok(expr)
                } else {
                    glux_error!(ParserError, "Expected ')'")
                }
            }

            // **處理一元運算子 `-5`、`+3`**
            Token::Operator(op) if op == "-" || op == "+" => {
                let op_clone = op.clone();
                self.advance();
                let right = self.parse_factor()?;
                Ok(ASTNode::BinaryOp {
                    op: op_clone,
                    left: Box::new(ASTNode::Number("0".to_string())), // `-x` 轉為 `0 - x`
                    right: Box::new(right),
                })
            }

            // **遇到意外的 token**
            token => glux_error!(ParserError, "Unexpected token in expression: {:?}", token),
        }
    }

    /// **解析函數呼叫（識別符後面有 `()`）**
    fn parse_function_call(&mut self, name: String) -> Result<ASTNode, GluxError> {
        self.advance(); // 跳過 `(`

        let mut args = Vec::new();
        if let Token::Symbol(')') = self.current_token() {
            self.advance(); // 跳過 `)`
            return Ok(ASTNode::Identifier(name)); // 暫時不處理函數參數
        }

        glux_error!(ParserError, "函數解析尚未實作")
    }
}

/// **外部解析函式：接收 Token 列表並返回 AST**
pub fn parse(tokens: Vec<Token>) -> Result<ASTNode, GluxError> {
    let mut parser = Parser::new(tokens);
    parser.parse()
}
